moz_security/irc.c

```diff
@@ -0,0 +1,550 @@
+/* -*- Mode: C; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
+/* vim: set ts=2 et sw=2 tw=80: */
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/** @brief Lua IRC integration @file */
+
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <sys/select.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <pthread.h>
+#include <netdb.h>
+#include <errno.h>
+#include <time.h>
+#include <openssl/ssl.h>
+#include <signal.h>
+
+#include "lauxlib.h"
+#include "lua.h"
+#include "lualib.h"
+
+#define MOZSVC_IRC_TABLE	"irc"
+#define MOZSVC_IRC		"mozsvc.irc"
+
+#define CHANQUEUE_LENGTH 32
+
+SSL_CTX *ctx;
+
+enum {
+  STATE_DOWN,
+  STATE_CONNECTING,
+  STATE_VERIFIED,
+  STATE_JOINED
+};
+
+struct queueent {
+  char            buf[400];
+  struct queueent *next;
+};
+
+struct ircconn {
+  pthread_t monitor;
+
+  char  *irc_nick;
+  char	*irc_chan;
+  char	*irc_chankey;
+  char	*irc_hn;
+  int   irc_port;
+
+  int     status; // STATE_DOWN, STATE_VERIFIED, etc...
+  time_t  rejoin_timer;
+
+  int	fd;
+  BIO	*bio;
+  SSL	*ssl;
+
+  char    inbuf[10240];
+  size_t  inbuf_len;
+
+  struct queueent *chanqueue;
+  int             chanqueue_size;
+  pthread_mutex_t chanqueue_lock;
+};
+
+struct args {
+  char **args;
+  int n;
+};
+
+static struct args *irclib_args_parse(char *);
+static void irclib_args_free(struct args *);
+static int irclib_parse(struct ircconn *, char *);
+void *irclib_monitor(void *);
+static int irclib_bio_write(BIO *, char *, size_t);
+static int irclib_tryconn(struct ircconn *);
+
+static int irclib_new(lua_State *);
+static int irclib_writeraw(lua_State *);
+static int irclib_writechan(lua_State *);
+static int irclib_status(lua_State *);
+int luaopen_irc(lua_State *);
```

Comment:
Other modules don't specify the function prototypes unless they are needed before the function definition. Also when provided the argument names are usually included.